File: src/main/java/de/fabilucius/advancedperks/core/database/DatabaseConfiguration.java
package de.fabilucius.advancedperks.core.database;

import de.fabilucius.advancedperks.core.configuration.Configuration;
import de.fabilucius.advancedperks.core.configuration.annotation.FilePathInJar;

import java.util.Arrays;

@FilePathInJar("database.yml")
public class DatabaseConfiguration extends Configuration {

    public String getFileLocation() {
        return this.getString("file.file_location", "data.db");
    }

    public SaveType getSaveType() {
        return Arrays.stream(SaveType.values())
                .filter(saveType -> saveType.name().equalsIgnoreCase(this.getString("save_type")))
                .findFirst()
                .orElse(SaveType.UNKNOWN);
    }

    public String getSqlUrl() {
        return this.getString("database.sql_url");
    }

    public Credentials getCredentials() {
        return new Credentials(
                this.getString("database.credentials.username"),
                this.getString("database.credentials.password")
        );
    }

}

File: src/main/kotlin/de/fabilucius/advancedperks/core/configuration/type/DatabaseConfiguration.kt
package de.fabilucius.advancedperks.core.configuration.type

import com.google.inject.Inject
import com.google.inject.Singleton
import com.google.inject.name.Named
import de.fabilucius.advancedperks.core.configuration.AbstractConfiguration
import de.fabilucius.advancedperks.core.configuration.annotation.FilePathInJar
import de.fabilucius.advancedperks.core.database.Credentials
import de.fabilucius.advancedperks.core.database.SaveType
import de.fabilucius.advancedperks.core.logging.APLogger
import java.io.File

@Singleton
@FilePathInJar("database.yml")
class DatabaseConfiguration @Inject constructor(@Named("configurationDirectory") configDir: File, logger: APLogger) :
    AbstractConfiguration(configDir, logger) {

    fun getFileLocation(): String = this.getString("file.file_location")!!

    fun getSqlUrl(): String = this.getString("database.sql_url")!!

    fun getCredentials(): Credentials = Credentials.withAuth(
        this.getString("database.credentials.username")!!,
        this.getString("database.credentials.password")!!
    )

    fun getSaveType(): SaveType = SaveType.entries.find {
        it.name.equals(getString("save_type")!!, true)
    } ?: SaveType.UNKNOWN

}